### PR TITLE
/health endpoint for K8 Liveness and Readiness probes

### LINKED ISSUE #3506: [REF-3134] K8 Liveness and Readiness probes
Expose a `/health` endpoint which reports the status of the reflex instance. This endpoint should check f.e. redis and postgres connections. This allows to configure liveness and readiness probes in k8 which prevents routing traffic to unhealthy reflex workers.

<sub>[REF-3134](https://linear.app/reflex-dev/issue/REF-3134/k8-liveness-and-readiness-probes)</sub>

### PR DESCRIPTION
**Pull request by samarth9201**
### All Submissions:

- [x] Have you followed the guidelines stated in [CONTRIBUTING.md](https://github.com/reflex-dev/reflex/blob/main/CONTRIBUTING.md) file?
- [x] Have you checked to ensure there aren't any other open [Pull Requests](https://github.com/reflex-dev/reflex/pulls ) for the desired changed?

### Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

### New Feature Submission:

- [x] Does your submission pass the tests? 
- [x] Have you linted your code locally prior to submission?

### Changes To Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully ran tests with your changes locally?

Created a '/health' endpoint which must be added as [reserved API Route](https://reflex.dev/docs/api-routes/overview/#r-e-s-e-r-v-e-d-%20-r-o-u-t-e-s). The /health endpoint checks the connectivity to Database and Redis. If either service is unavailable or unhealthy, the endpoint returns a 503 HTTP status, and 200 HTTP status if everything is alright (Ref : [Configure Liveness, Readiness and Startup Probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#:~:text=To%20perform%20a%20probe%2C%20the%20kubelet%20sends,kubelet%20kills%20the%20container%20and%20restarts%20it.)). Checking more services can be added to this if required.

**Approach**

1. Created the /health endpoint asynchronous function that checks the status of database and Redis connections. It initializes a health status dictionary with default values indicating that both services are healthy.
2. For Database, it attempts to connect to the database and execute a simple query (SELECT 1). If the connection fails or an error is raised, the health status for the database is updated to reflect the failure.
3. For Redis, it tries to ping the Redis server. If the ping fails, the health status for Redis is updated to indicate a failure.
4. The endpoint returns a JSON response containing the health status of both services along with an appropriate HTTP status code (200 for healthy, 503 for unhealthy).

closes #3506

### CODE REVIEW

File: reflex/app.py [hunk 1]
```diff
@@ -1535,3 +1564,12 @@ async def on_ping(self, sid):
...
         await self.emit(str(constants.SocketEvent.PING), "pong", to=sid)
+
+    async def on_health(self, sid):
+        """Event for testing the Health of Reflex instance.
+
+        Args:
+            sid: The Socket.IO session id.
+        """
+        # Emit the test event.
+        await self.emit(str(constants.SocketEvent.HEALTH), "ok", to=sid)
```
Comment (masenf) [inline reflex/app.py:1575]: i don't think we need a websocket event handler as it's essentially equivalent to what we already have here for `ping`

File: reflex/app.py [hunk 2]
```diff
@@ -1319,6 +1320,34 @@ async def ping() -> str:
     return "pong"
 
 
+async def health():
```
Comment (masenf) [inline reflex/app.py:1323]:
```suggestion
async def health() -> JSONResponse:
```

File: reflex/utils/prerequisites.py [hunk 3]
```diff
@@ -344,6 +345,29 @@ def parse_redis_url() -> str | dict | None:
     return dict(host=redis_url, port=int(redis_port), db=0)
 
 
+def get_redis_status() -> bool | None:
+    """Checks the status of the Redis connection.
+
+    Attempts to connect to Redis and send a ping command to verify connectivity.
+
+    Returns:
+        int or bool or str: The status of the Redis connection:
```
Comment (masenf) [inline reflex/utils/prerequisites.py:354]: the function appears to only return a `bool` or `None`, and the docstring doesn't describe the `None` case

File: reflex/constants/event.py [hunk 4]
```diff
@@ -11,6 +11,7 @@ class Endpoint(Enum):
     EVENT = "_event"
     UPLOAD = "_upload"
     AUTH_CODESPACE = "auth-codespace"
+    HEALTH = "health"
```
Comment (masenf) [inline reflex/constants/event.py:14]:
```suggestion
    HEALTH = "_health"
```

recommend a leading underscore for this endpoint

**Review by masenf — COMMENTED** [review]
mostly looks good, really like the tests. a couple of feedbacks

File: reflex/app.py (hunk 1 above)
Comment (samarth9201) [reply reflex/app.py]: Yep, making the change

File: reflex/utils/prerequisites.py (hunk 3 above)
Comment (samarth9201) [reply reflex/utils/prerequisites.py]: Updated the docstring

File: reflex/constants/event.py [hunk 5]
```diff
@@ -47,6 +48,7 @@ class SocketEvent(SimpleNamespace):
 
     PING = "ping"
     EVENT = "event"
+    HEALTH = "health"
```
Comment (masenf) [inline reflex/constants/event.py:51]: not needed with the socket event removed

Comment (samarth9201) [reply reflex/constants/event.py]: Fixed it

File: reflex/model.py [hunk 6]
```diff
@@ -51,6 +51,26 @@ def get_engine(url: str | None = None) -> sqlalchemy.engine.Engine:
...
+        bool: The status of the database connection:
+            - True: The database is accessible.
+            - False: The database is not accessible.
+    """
+    status = True
+    try:
+        engine = get_engine()
+        with engine.connect() as connection:
+            connection.execute(sqlalchemy.text("SELECT 1"))
+    except Exception:
```
Comment (benedikt-bartscher) [inline reflex/model.py:69]: Can we narrow this Exception down?

**Review by benedikt-bartscher — APPROVED** [review]
Thank you, @samarth9201, LGTM!
We could make `get_redis_status` and `get_db_status` async and run them at the same time to slightly improve performance

File: reflex/model.py (hunk 6 above)
Comment (samarth9201) [reply reflex/model.py:69]: Yes, we can narrow the exception handling to sqlalchemy.exc.OperationalError to specifically address connectivity issues, which aligns with the function's goal of checking database availability. However, other exceptions like ArgumentError or NoModuleError could occur due to misconfiguration or missing dependencies. To handle these cases, we could use a general except Exception block as a fallback to catch any unforeseen errors and include logging to help with debugging. Does this approach sound good?

**Comment by samarth9201** [comment]
Hi @masenf, could you please review the changes and approve the PR if everything looks good? Thank you!

**Review by masenf — APPROVED** [review]
awesome thanks for your contribution and patience 🙏

**Pull request merged**